OpenTrons/opentrons-api
# fix(robot-server): Refactor hardware wrapper and other lifetime dependencies

# Overview

This PR refactors and fixes how robot-server sets up and tears down its *lifetime dependencies*. Lifetime dependencies are dependencies that should be initialized when the server starts, and cleaned up when the server stops.

The goal is to address the specific issues of #8051 and #8023, and also some of the structural traps (in my opinion) that allowed them to exist in the first place.

# Changes

## Fix #8051 and fix #8023

Before this PR, all of our dependencies were created through FastAPI dependency functions.

Some of those dependency objects needed special initialization or cleanup. To do this, our startup/shutdown code fetched the objects like this:

https://github.com/Opentrons/opentrons/blob/4be5bc00b7db5958238e53ad7738af31a046fca0/robot-server/robot_server/app.py#L58-L63

(`get_hardware_wrapper()` was one of our FastAPI dependency functions.)

It turns out that this is was invalid thing to do, and could corrupt objects in subtle ways. See #8051 for details. We needed a way for our startup/shutdown code to do its job without calling any FastAPI dependency functions.

This PR accomplishes that by flipping the relationship between our startup code and FastAPI dependency functions.

* Our startup code now initializes lifetime dependencies on its own. It places the result objects in a dedicated place.
* While the server is running, our FastAPI dependency functions act as simple getters, fetching from that dedicated place.
* Our shutdown code fetches the lifetime dependency objects from that dedicated place and runs each one's cleanup code.

## Separate our initialization logic from the machinery that runs that logic in the background

Before this PR, we had a `HardwareWrapper` class that acted as two things:

1. A factory function. It coordinated things to initialize the hardware API however we needed it to be initialized.
2. A handle for doing that in the background. It let us start initializing as soon as the server started, allowing other things to happen concurrently. `HardwareWrapper` would own the background task and provided access to the hardware API whenever it became available.

This PR separates those concerns into two independent units.

The first new unit is a `hardware_initialization` module. This has an `async def initialize()` function with a simple interface: it either returns a fully-formed, fully-usable hardware object, or it raises an exception.

The second unit is a `slow_initializing` module. This has a wrapper class that wraps any factory function, runs it in the background, and provides access to its result whenever it becomes available.

Separating these units and fleshing them out wins us:

* The ability to test and document these two concerns independently, and hopefully more thoroughly.
* A simplified hardware API object, with fewer windows for misuse. The old `HardwareWrapper` exposed initialization methods to endpoint-level code, whereas now we just expose the fully-formed, ready-to-use result. We also remove windows for partial initialization, simplifying the number of possible states that the object could be in, hopefully making it easier to reason about.
* The ability for endpoint-level code to distinguish between the hardware API having failed to initialize, versus simply not being ready yet. The old `HardwareWrapper` conflated the two.
* No orphaned background tasks. The old `HardwareWrapper` orphaned its background task, so exceptions wouldn't propagate anywhere, among other potential problems.

Lifetime dependencies other than the hardware API also now use the `slow_initializing` machinery. This has to do with the changes explained in the section above. Formerly, if dependency `get_foo()` depended on `get_hardware()`, and something tried accessing `get_foo()` before `get_hardware()` was ready, FastAPI's dependency resolution magic would make `get_foo()` automatically raise a "hardware not ready" exception. But now, since we're initializing lifetime dependencies outside of FastAPI, we have to implement that ourselves.

## Set up a pattern of defining lifetime dependencies as context managers


* It's natural to express each individual lifetime dependency as a context manager—it's a resource that needs cleanup, and context managers are Python's way of doing resources that need cleanup.
    * Doing this as a pattern also made it more obvious that we were forgetting to clean up certain lifetime dependencies, so that's fixed now.
* We use standard Python tools to compose our many small context managers into one big monolithic one. The monolithic one returns a `dataclass` containing every lifetime dependency.
    * This simplifies the module's interface: it stashes one big object for external code to retrieve later, instead of many small objects.
    * It's also convenient to transform this into a startup/shutdown function pair.
* FastAPI and Starlette seem to be figuring out how to support lifetime dependencies in a more integrated way. The exact API still seems up in the air, but, according to some GitHub threads that I've linked in the code, it seems very very likely that it will somehow be based on context managers or `@contextlib.contextmanager`-like factories. So, if all of this is just a holdover until FastAPI adds proper support, our holdover will at least be structurally similar to the ultimate solution.


# Review requests

## Code review requests

* After getting a feel for the production code, please look over the changes to tests and test fixtures. Those are the parts I'm least confident about.

## Manual test requests

* Please run `make test` and make sure it doesn't hang for you. At some of the intermediate commits in this PR, I thought I was experiencing mysterious hanging tests, but I can't reproduce it now.
* If you have a real OT-2 handy...
    * Please push to it and do a basic smoke test like homing or running a protocol. I've already done this with PE feature flags disabled, so it would be helpful if someone else did it with the feature flags enabled.
    * If you want: SSH in; tail logs with `journalctl -f opentrons-robot-server`; run `systemctl stop opentrons-robot-server`; and make sure the service stops cleanly, except for the expected errors listed below.
* If you want, run `make dev` and make sure...
    * You can still connect to the virtual robot.
    * The dev server exits promptly and cleanly when you interrupt with `^C`, except for the expected errors listed below.
* If you want, run through the repro steps in #8023 and make sure that it is indeed fixed.

I will also do all of these before merging.

### Expected errors

* Anything relating to closing a WebSocket connection (like `Cannot call "send" once a close message has been sent`). I don't know exactly what this is, but it's not new. See #8155.
* `Task was destroyed but it is pending!` errors relating to `opentrons/hardware_control/poller.py`. This is https://github.com/Opentrons/opentrons/issues/8239.
* On a real robot, server shutdown hanging on something like "waiting for background tasks to finish." I can't reproduce this, but I don't think it's new, and I don't think it could be related to this PR because this PR registers no background tasks with Starlette.

# Risk assessment

Medium.

Risk is brought up because this is critical to everything that the OT-2 does.

Risk is brought down because problems with this code *should* be obvious with a smoke test.
